doc/developer/design/20230720_single_statement_explitic_transaction.md

```diff
@@ -0,0 +1,105 @@
+
+Each `Session` has a `tranasction` property of type `enum TransactionStatus`.
+It is the responsibility of the various protocol handlers (pgwire, http, ws) to correctly manage this.
+A `TransactionStatus` is an enum with variants:
+
+- `Default`
+- `Started`
+- `InTransaction`
+- `InTransactionImplicit`
+- `Failed`
+
+`Default` is the initial state, and also the state when there is no in-progress transaction.
+In order for a session to execute a statement, the protocol handler must set this to one of the other states then end the transaction with a commit or rollback.
+`Started` is used for a single statement, so an immediate and implied commit follows.
+`InTransaction` is used when a user explicitly types `BEGIN`, and can only be exited when a user sends `COMMIT` or `ROLLBACK`.
+`InTransactionImplicit` is used when a user sends multiple statements in the same query (`SELECT 1; SELECT 2`), and an implicit `BEGIN` and `COMMIT` wrap it.
+`Failed` can only be entered when a statement in a transaction has error'd from the `InTransaction` state.
+If failures occur in the other in-progress states, an implicit `ROLLBACK` must be sent by the handler.
+
+All status except `Default` have an inner `Transaction` object that tracks what operations have happened in a transaction determined by the first statement.
+These operations are: `Peeks`, `Subscribe`, `Writes`, `None`.
+`None` is the default and can transition to any other operation.
+The other operations can be merged with like operations, except `Subscribe` which supports only a single `SUBSCRIBE`.
+
+The `ReadyForQuery` message is sent by the server when it is ready to receive new queries, and it includes the current transaction status code: `Idle`, `InTransaction`, or `Failed`.
+`InTransaction` and `Failed` correspond to the similarly named variants above and can only appear in explicit transactions.
+`Idle` is return in the `Default` state, because all other transaction states implicitly close and should never be active when deciding the status code.
+
+The pgwire, http, and ws handlers enforce these requirements and state changes.
+
+#### Changes from this design
+
+This design adds a new operation to the inner transaction: `SingleStatement`.
+This operation records a single statement that is later retrieved and executed during `COMMIT`.
+
+This design also adds what is essentially a new handler: the coordinator itself.
+When a transaction is being committed and its inner operation is `SingleStatement`, the inner statement is queued in the coordinator for execution by a new method that is serving as a handler: `sequence_execute_single_statement_transaction`.
+This method:
+1. Ensures the session's transaction is in `Default`, which should be a side effect of running it through `sequence_end_transaction` from the `COMMIT`.
+2. Puts the session's transaction into `Started` (single-statement, implicit transaction).
```

Comment:
Similarly it was a bit confusing to me that we set the state to an implicit transaction when the transaction is an explicit transaction. It sounds like we're just trying to take advantage of some nice properties that implicit transactions have but it's not really an implicit transaction? 

If so I think it would be clearer to add a new state that has those properties but is specific to these single statement transactions. Otherwise whenever someone makes changes to the implicit transaction state they'll have to remember that they're also making changes to the single statement transaction state, and it seems easy to forget about this.

Reply:
I'm not sure doing this helps us much. We would still need to teach sequence_end_transaction to detect this status and it'd then still have to enqueue a message to do the work, and the only way to do the work is to call handle_execute, which is the thing doing all the transaction status checking, forcing the manual starting of an implicit transaction. We could refactor handle_execute to split up those parts and skip the checking, but I don't think it's safer.

Also, note way above I incorrectly used `InTransactionImplicit` (multi statement implicit) where it should have been `Started` (single statement implicit). `Started` is pretty clear about what's going on, and seems like a safe status to use.

I think my summary understanding of your idea (a new status) is: it's not clear to me how we would use that status to both delay execution and then later actually execute.

Reply:
Done.